src/ucs/sys/topo/base/topo.c

```diff
@@ -51,32 +52,73 @@ const ucs_sys_dev_distance_t ucs_topo_default_distance = {
 static ucs_topo_global_ctx_t ucs_topo_global_ctx;
 
 
-static ucs_bus_id_bit_rep_t
-ucs_topo_get_bus_id_bit_repr(const ucs_sys_bus_id_t *bus_id)
+/* Global list of topology detectors */
+UCS_LIST_HEAD(ucs_sys_topo_methods_list);
+
+
+ucs_sys_topo_method_t *
+ucs_sys_topo_find_method(const char *name)
 {
-    return (((uint64_t)bus_id->domain << 24) |
-            ((uint64_t)bus_id->bus << 16)    |
-            ((uint64_t)bus_id->slot << 8)    |
-            (bus_id->function));
+    ucs_sys_topo_method_t *method;
+
+    ucs_list_for_each(method, &ucs_sys_topo_methods_list, list) {
+        if (!strcmp(name, method->name)) {
+            return method;
+        }
+    }
+
+    return NULL;
 }
 
-void ucs_topo_init()
+static ucs_sys_topo_method_t *ucs_sys_topo_get_method()
 {
-    ucs_spinlock_init(&ucs_topo_global_ctx.lock, 0);
-    kh_init_inplace(bus_to_sys_dev, &ucs_topo_global_ctx.bus_to_sys_dev_hash);
-    ucs_topo_global_ctx.num_devices = 0;
+    ucs_sys_topo_method_t *method = NULL;
+    unsigned i;
+
+    for (i = 0;
+         (i < ucs_global_opts.topo_methods.count) && (method == NULL);
+         ++i) {
+        method = ucs_sys_topo_find_method(ucs_global_opts.topo_methods.names[i]);
```

Comment:
seems ucs_sys_topo_find_method() is not needed: can just have nested loop here, and if strcmp returns 0 - return method

also, maybe cache the selected method in static var?